crates/bitcoind_rpc/src/lib.rs

```diff
@@ -0,0 +1,243 @@
+//! mempool.
+#![warn(missing_docs)]
+
+use std::collections::BTreeMap;
+
+use bitcoin::{block::Header, Block, BlockHash, Transaction};
+pub use bitcoincore_rpc;
+use bitcoincore_rpc::bitcoincore_rpc_json;
+
+/// A structure that emits data sourced from [`bitcoincore_rpc::Client`].
+///
+/// Refer to [module-level documentation] for more.
+///
+/// [module-level documentation]: crate
+pub struct Emitter<'c, C> {
+    client: &'c C,
+    start_height: u32,
+
+    emitted_blocks: BTreeMap<u32, BlockHash>,
+    last_block: Option<bitcoincore_rpc_json::GetBlockResult>,
+
+    /// The latest first-seen epoch of emitted mempool transactions. This is used to determine
+    /// whether a mempool transaction is already emitted.
+    last_mempool_time: usize,
+
+    /// The last emitted block during our last mempool emission. This is used to determine whether
+    /// there has been a reorg since our last mempool emission.
+    last_mempool_tip: Option<(u32, BlockHash)>,
+}
+
+impl<'c, C: bitcoincore_rpc::RpcApi> Emitter<'c, C> {
+    /// Constructs a new [`Emitter`] with the provided [`bitcoincore_rpc::Client`].
+    ///
+    /// `start_height` is the block height to start emitting blocks from.
+    pub fn new(client: &'c C, start_height: u32) -> Self {
+        Self {
+            client,
+            start_height,
+            emitted_blocks: BTreeMap::new(),
+            last_block: None,
+            last_mempool_time: 0,
+            last_mempool_tip: None,
+        }
+    }
+
+    /// Emit mempool transactions, alongside their first-seen unix timestamps.
+    ///
+    /// Ideally, this method would only emit the same transaction once. However, if the receiver
+    /// filters transactions based on whether it alters the output set of tracked script pubkeys,
+    /// there are situations where we would want to re-emit. For example, if an emitted mempool
+    /// transaction spends a tracked UTXO which is confirmed at height `h`, but the receiver has
+    /// only seen up to block of height `h-1`, we want to re-emit this transaction until the
+    /// receiver has seen the block at height `h`.
+    ///
+    /// In other words, we want to re-emit a transaction if we cannot guarantee it's ancestors are
+    /// already emitted.
+    pub fn mempool(&mut self) -> Result<Vec<(Transaction, u64)>, bitcoincore_rpc::Error> {
+        let client = self.client;
+
+        let prev_mempool_tip = match self.last_mempool_tip {
+            // use 'avoid-re-emission' logic if there is no reorg
+            Some((height, hash)) if self.emitted_blocks.get(&height) == Some(&hash) => height,
+            _ => 0,
+        };
+
+        let prev_mempool_time = self.last_mempool_time;
+        let mut latest_time = prev_mempool_time;
+
+        let txs_to_emit = client
+            .get_raw_mempool_verbose()?
+            .into_iter()
+            .filter_map({
+                let latest_time = &mut latest_time;
```

Comment:
It seems to me that this line was put here just to please the compiler, but actually it works fine even if you remove it. Unless there's some logic difference that I don't see.

Reply:
last_seen will be copied into the closure and the original last_seen will stay the same.

Reply:
> The original last_seen will stay the same.

Is it? The new latest_time is just a mutable reference to the original latest_time, so you're modifying the original as well inside the closure.

But if I'm wrong, and in the closure you have a different latest_time that you're modifying only inside the closure, why is that needed? In the closure you only modify latest_time, you never read it.

Reply:
Sorry I didn't try, removing the line some tests fail, so it is needed. I just don't get why :(

Reply:
latest_time implements copy, so it won't be moved into the closure, it'll get copied.